zombies/src/main/java/com/github/phantazmnetwork/zombies/game/map/ZombiesMap.java

```diff
@@ -0,0 +1,224 @@
+package com.github.phantazmnetwork.zombies.game.map;
+
+import com.github.phantazmnetwork.commons.Namespaces;
+import com.github.phantazmnetwork.commons.component.ComponentBuilder;
+import com.github.phantazmnetwork.commons.component.ComponentException;
+import com.github.phantazmnetwork.commons.component.DependencyProvider;
+import com.github.phantazmnetwork.commons.component.annotation.ComponentDependency;
+import com.github.phantazmnetwork.core.ClientBlockHandler;
+import com.github.phantazmnetwork.commons.Tickable;
+import com.github.phantazmnetwork.commons.vector.Region3I;
+import com.github.phantazmnetwork.commons.vector.Vec3D;
+import com.github.phantazmnetwork.commons.vector.Vec3I;
+import com.github.phantazmnetwork.mob.spawner.MobSpawner;
+import com.github.phantazmnetwork.zombies.game.SpawnDistributor;
+import com.github.phantazmnetwork.zombies.map.*;
+import net.kyori.adventure.key.Key;
+import net.minestom.server.instance.Instance;
+import net.minestom.server.instance.block.Block;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.UnmodifiableView;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+public class ZombiesMap extends PositionalMapObject<MapInfo> implements Tickable {
+    @ComponentDependency("zombies.dependency.map.context")
+    public record ObjectContext(@NotNull Instance instance,
+                                @NotNull MobSpawner spawner,
+                                @NotNull ClientBlockHandler blockHandler,
+                                @NotNull SpawnDistributor spawnDistributor) {
+        public static final Key DEPENDENCY_KEY = Key.key(Namespaces.PHANTAZM, "zombies.dependency.map.context");
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ZombiesMap.class);
+
+    private final List<Room> unmodifiableRooms;
+    private final List<Door> unmodifiableDoors;
+    private final List<Window> unmodifiableWindows;
+    private final List<Spawnpoint> unmodifiableSpawnpoints;
+    private final List<Round> unmodifiableRounds;
+
+    private int roundIndex = 0;
+    private Round currentRound;
+
+    /**
+     * Constructs a new instance of this class.
+     *
+     * @param info     the backing data object
+     * @param instance the instance which this MapObject is in
+     */
+    public ZombiesMap(@NotNull MapInfo info, @NotNull ComponentBuilder builder, @NotNull Instance instance,
+                      @NotNull MobSpawner mobSpawner, @NotNull ClientBlockHandler blockHandler,
+                      @NotNull SpawnDistributor spawnDistributor) {
+        super(info, info.settings().origin(), instance);
+
+        ObjectContext context = new ObjectContext(instance, mobSpawner, blockHandler, spawnDistributor);
+        DependencyProvider provider = DependencyProvider.ofDependencies(context);
```

Comment:
I'm a little confused - why use a provider here?

Reply:
subclasses of Action<T> might want to have access to dependencies like the map's SpawnDistributor, among other things

Reply:
these would all be annotated or Keyed?

Reply:
nvm it's the actual context that is passed in